SalesforceCommerceCloud/pwa-kit
# [Extensibilily]PWA ProductView Extensbility POC

# Problem

When a pwa-kit project is generated through the generator, the partner developers have no means to consume our update (e.g new features or bug fixes) in the Retail React App template programmatically. 

# Solution
1. Modulize some of our mid-size components, and re-design them to allow a wide range of layout flexibility. By doing this, developers can have the flexibility of customizing without rebuilding the layout of the component. The only time they might need to do that is when they don't want to use our components, yet build their own.  

2. Modulize pages business logic into hooks for consumption and upgradability in their own React components

Whether these two will be in the same package is up for debate


# Description
This PR demonstrates the solution by modularizing ProductView into a different package (I called it `pwa-kit-ecom`)  and allowing the Retail React App template to consume those components in the PDP. The ProductView is designed in a way that allows users to customize part of it for children. This would allow a wide range of flexibility to users whether they want to use it as is or partially replace or fully replace ProductView in PDP completely. 

The PR also extracts the business of ProductView into a separate hook. (useProduct.js), this allows partners to reuse the hook in case they don't like how ProductView is designed.

This demo is to demonstrate the possibility of customisability and upgradability, it has not taken into the roadblocks of other factors like localization, multi-site, etc

# Why this approach
1. It allows developers that generated a pwakit project to consume update from time to time if and only if they choose use our ProductView. If they choose to replace the component, they can't consume the change. However, it is possible for them to consume hook containing ProductView component. 
2. It allows us to roll out small changes, moving a groups of hooks and components, and able to gather some feedback along the way. It is a great way to test whether this method might benefit partners in a long run
3. The key here is not to move everything components but to select the ones that can save developers from rebuilding a page UI layout. 

# How to Test-Drive This PR

- check out this branch
- `npm ci` at root
- `npm start` at Retail app
- Go to PDP and see three flexbility that offers from ProductView modularization
